Update IAR Armv8M changes

deepikabhavnani opened the pull request:
### Description
Update IAR flags for Armv8M devices.

    Cortex-M23 / Cortex-M33 CPU settings for baseline and mainline
    profile (with optional floating and dsp options) updated.

Dependent on https://github.com/ARMmbed/mbed-os/pull/9480
### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

Commit:
Update IAR flags for Armv8M devices.

Cortex-M23 / Cortex-M33 CPU settings for baseline and mainline
profile (with optional floating and dsp options) updated.

Commit:
M2351: Update code to support IAR 8.x builds

ciarmcom commented:
@deepikabhavnani, thank you for your changes.
@ARMmbed/mbed-os-tools @ARMmbed/mbed-os-maintainers please review.

kjbracey commented on tools/toolchains/iar.py:
```diff
@@ -74,12 +76,15 @@ def __init__(self, target, notify=None, macros=None, build_profile=None,
-        elif target.core == "Cortex-M23" or target.core == "Cortex-M33" or target.core == "Cortex-M33F":
-            self.flags["asm"] += ["--cmse"]
-            self.flags["common"] += ["--cmse"]
+        elif target.core.startswith("Cortex-M33F"):
```
Is this whole section being more complicated than necessary? Current IAR manual says it understands `Cortex-M33.fp`, `Cortex-M7.fp.sp` and `Cortex-M7.fp.dp` as cpu options, so you may not need to be passing separate `--fpu` options at all.

deepikabhavnani replied:
Agree, but will leave the ‘fpu’ setting cleanup for another PR. Will have to also consider if fpu is default enabled or not in the core selection. As for Cortex-M33 selection as core fpu and dsp is enabled as default and we need to set .no_dsp instead

kjbracey replied:
Ah, I see you going for "no_fp" here: https://github.com/ARMmbed/mbed-os/pull/9433/files.

<strike>I don't think that's right - will comment there.</strike>  Ah, that was for arm tools.

kjbracey commented on tools/toolchains/iar.py:
```diff
@@ -74,12 +76,15 @@ def __init__(self, target, notify=None, macros=None, build_profile=None,
 
         # Create Secure library
-        if target.core == "Cortex-M23" or self.target.core == "Cortex-M33" or self.target.core == "Cortex-M33F":
+        if ((target.core.startswith("Cortex-M23") or target.core.startswith("Cortex-M33"))
```
Could this be an architecture >=8 check (CORE_ARCH from tools.targets)?

(This is going to need revisiting at some point - shouldn't just assume that a v8 chip or M33 is going to be using separate non-secure and secure images).

deepikabhavnani replied:
I am not sure if we have a way to check arch @theotherjimmy please suggest

kjbracey replied:
That would be the CORE_ARCH thing from tools.targets - was used in this commit I'm reverting: #9398

theotherjimmy replied:
What @kjbracey-arm said ^. That's the way to check arch version.

deepikabhavnani replied:
Done 👍

kjbracey commented on tools/toolchains/iar.py:
```diff
@@ -47,10 +47,12 @@ def __init__(self, target, notify=None, macros=None, build_profile=None,
             cpuchoice = "Cortex-M7"
-        elif target.core.startswith("Cortex-M23"):
-            cpuchoice = "8-M.baseline"
+        elif target.core.startswith("Cortex-M33FD"):
```
It's not being introduced by this PR, but I'm not a huge fan of this "D" meaning something different to the "D" in "M7FD".

In the past ARM used "E" to indicate the DSP extension, so M33FE might make more sense?

Although I'd be inclined to start following the modern tools' lead and go with "Cortex-M33.no_dsp.no_fp" - pick either the armcc or gcc form to use for Mbed OS, rather than making up our own single-letter suffixes.

deepikabhavnani replied:
@kjbracey-arm - Agree to you on this, but I will leave this out of the scope of this PR.

@SenRamakri - It will be good to add this as task in tools. It will impact all M devices with different variants and all toolchians, part of tools cleanup

Commit:
Update `cmse` flag based on core_arch instead of different cores

Commit:
Correct the floating+dsp options for Cortex-M processors

As per the IAR Development guide, below options for CPU are valid

1. Cortex-M33
2. Cortex-M33.no_dsp (core without integer DSP extension)
3. Cortex-M33.fp (floating-point unit with support for single precision)
4. Cortex-M33.no_se (core without support for TrustZone)
5. Cortex-M4
6. Cortex-M4F
7. Cortex-M7
8. Cortex-M7.fp.dp (floating-point unit with support for double precision)
9. Cortex-M7.fp.sp (floating-point unit with support for single precision)

deepikabhavnani commented:
@kjbracey-arm @theotherjimmy - Please review

theotherjimmy reviewed (approved):
Looks good supposing that we're leaving the core-reworking to another PR.

deepikabhavnani commented:
> supposing that we're leaving the core-reworking to another PR.

Yes core reworking will in another PR (by tools team :-) )

cyliangtw reviewed (changes requested):
In iar_definitions.json, M2351 part should be:
"M2351KIAAEES": {
"OGChipSelectEditMenu": "M2351 series\tNuvoton M2351 series"
}

Commit:
Set the DSP option as `E` for Cortex-M33

bridadan commented on tools/toolchains/iar.py:
```diff
@@ -68,20 +80,6 @@ def __init__(self, target, notify=None, macros=None, build_profile=None,
         ]
-        if target.core == "Cortex-M7FD":
-            asm_flags_cmd += ["--fpu", "VFPv5"]
-            c_flags_cmd.append("--fpu=VFPv5")
```
is anything lost by not specifying the `--fpu` in either the assembler or compiler?

deepikabhavnani replied:
In case of IAR no, see https://github.com/ARMmbed/mbed-os/pull/9431#discussion_r249349584

deepikabhavnani commented:
Rebased to pull in dependent PR changes

bridadan commented on tools/toolchains/iar.py:
```diff
@@ -43,23 +44,34 @@ def check_executable():
-            cpuchoice = "8-M.mainline"
-        else:
-            cpuchoice = target.core
+        mbedToolchain.__init__(self, target, notify, macros, build_dir=build_dir,build_profile=build_profile)
```
style nit: we usually have spaces after commas
```suggestion
        mbedToolchain.__init__(self, target, notify, macros, build_dir=build_dir, build_profile=build_profile)
```

bridadan reviewed (approved):
Nice work! One non-blocking style nit pick.

bridadan committed:
Add missing space - coding style nit

Co-Authored-By: deepikabhavnani <deepika.bhavnani@arm.com>

cmonr commented:
CI started
